### DISCUSSION TITLE
How to iterate over pydantic's Field type.

### DISCUSSION THREAD

**Discussion by ghost** [pydantic/pydantic#4903] [Question]
Hello everyone, I'm fairly new to pydantic and tried to loop over a Field(default_factory= dict) property from a DTO of mine, but got FieldInfo object is not iterable.
I use the DTO to instatiate some class properties inside the constructor, and one of them is a Dict[str, int]. 

DTO:

```
from typing import Dict, Optional
from pydantic.dataclasses import dataclass
from pydantic import Field
from .DTOs.bar_dto import BarDto

@dataclass
class FooDto(BarDto):
    description: Optional[str]= ''
    baz_dict: Dict= Field(default_factory= dict)

```

Class:

```
class Mock(MockBase):
    def __init__(self, dto_: FooDto):
        ...
        self.baz_dict= dto_.baz_dict
    ...

```

The error comes when I try doing:

```
dict(
    self.baz_dict,
    ...
)
```

How can I iterate over the value I give the baz_dict Field, if I wish to do so?

PS.: Sorry if I wasn't clear enough, feel free to ask anything, I'll try to answer them to the best of my abilities.

**Comment by samuelcolvin** [comment]
I'm not sure what you mean by "DTO"?

I think you need to use `MyModel.__fields__`, you can then access `default_factory`:

https://github.com/pydantic/pydantic/blob/e8934a53eba7573488f6929ca6a5c2a6d411cfdd/pydantic/fields.py#L153

and `default`:

https://github.com/pydantic/pydantic/blob/e8934a53eba7573488f6929ca6a5c2a6d411cfdd/pydantic/fields.py#L152

**Reply by ghost** [reply]
DTO as in "Data Transfer Object", I wasn't using BaseModel, I had forgotten to include pydantic.dataclasses dataclass, which is what I'm using

**Comment by mkdir700** [comment]
I guess this is your code, like this:

```python
from typing import Dict, Optional

from pydantic import Field


class FooDto(object):  # incorrect
    description: Optional[str] = ""
    baz_dict: Dict = Field(default_factory=dict)


class Mock:

    def __init__(self, dto: FooDto):
        self.baz_dict = dto.baz_dict

    def dict(self):
        return dict(self.baz_dict)


print(Mock(FooDto()).dict())
```

run above code, its raise this error:

```
TypeError: 'FieldInfo' object is not iterable
```

cause: DTO class is not subclass of `BaseModel`

**Reply by samuelcolvin** [reply]
`FieldInfo` is just a class, inspect it like any other

https://github.com/pydantic/pydantic/blob/63c683d9fc7b494bf81e964f97ec2f9b78b8e09d/pydantic/fields.py#L97

You might want to use `field.__slots__` or event something like `{s: getattr(field, s) for s in field.__slots__}`

**Reply by ghost** [reply]
I'll check this out. I actually tried looking for a way to access the properties on this FieldInfo class, but didn't find what I wanted. I'll try your solution.

PS.: Updated my question, forgot to include pydantic's dataclass, which is what I'm using.

**Reply by samuelcolvin** [reply]
You want to access `MyDatacalsss.__pydantic_model__.__fields__`.

**Reply by ghost** [reply]
I did this, and was able to access the default_factory from the field, but I can't seem to access the data stored in this dict:

`logging.info(getattr(self.baz_dict, 'default_factory'))` only logs `<class 'dict'>`

If I do `self.baz_dict.default_factory` I get the same output as above.

How can I access the data stored inside this dict? I'm trying but can't seem to get it right.

And when I access the __slots__ property, I get a tuple containing strings for the properties names only as defined in the source code, not the key/value pairs.

**Reply by samuelcolvin** [reply]
> And when I access the slots property, I get a tuple containing strings for the properties names only as defined in the source code, not the key/value pairs.

Hence why I added `{s: getattr(field, s) for s in field.__slots__}`.

To be clear - non of this is specific to pydantic, it just general introspection with python. There's a limit to how much help I can really provide at this point. You might try stackoverflow for general questions like this.

**Reply by ghost** [reply]
@samuelcolvin yeah, I get that, I've used this dict comprehension before.
Thanks a lot for the help, I got it working, what I was missing was the dataclass decorator, it was just a silly error with an error message I didn't quite understand. Anyways, learned quite a bit in this discussion, so thanks for the help everyone.

PS.: I believe that since I didn't decorate the class with dataclass, the dict I was trying to attribute to the Field method was not working, so what was being read was a `dict` type
